### PR TITLE
add a trait for really commmon base fields

### PR DESCRIPTION
**Pull request by dawehner**

### CODE REVIEW

**Comment by bojanz** [comment]
This imposes the convention of using "id" as the name of the id field, which neither core nor contrib ever followed (it was always product_id, group_id, wehner_id).

It also adds a langcode without a default widget, "language_select" is usually set as one?

Ultimately I'm unsure how useful this trait actually is.

**Comment by fago** [comment]
> This imposes the convention of using "id" as the name of the id field, which neither core nor contrib ever followed (it was always product_id, group_id, wehner_id).

Right, but the advantage of "easy joins" is an argument of the past, we are not writing them manually anyway. So why not be consistent with the method name and use "id".

File: src/EntityBaseFieldsTrait.php [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+ * Contains \Drupal\entity\EntityBaseFieldsTrait.
+ */
+
+namespace Drupal\entity;
+
+use Drupal\Core\Field\BaseFieldDefinition;
+
+trait EntityBaseFieldsTrait {
+
+  protected function entityBaseFieldsDefaultBaseFields() {
```
Comment (dawehner) [inline src/EntityBaseFieldsTrait.php]: - add the bundle, just config entity refs
- <del>add the label</del>

File: src/EntityBaseFieldsTrait.php [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+/**
+ * @file
+ * Contains \Drupal\entity\EntityBaseFieldsTrait.
+ */
+
+namespace Drupal\entity;
+
+use Drupal\Core\Field\BaseFieldDefinition;
+
+trait EntityBaseFieldsTrait {
```
Comment (dawehner) [inline src/EntityBaseFieldsTrait.php]: EntityKeysFieldsTrait

File: src/EntityBaseFieldsTrait.php (hunk 1 above)
Comment (dawehner) [inline src/EntityBaseFieldsTrait.php]: entityKeysBaseFieldDefinitions

Comment (dawehner) [inline src/EntityBaseFieldsTrait.php]: use t('Label') + type as field name for bundles

**Comment by dawehner** [comment]
Thank you for the great discussion! Addressed stuff now

**Comment by bojanz** [comment]
$entity_type->getKey('id') and others can be NULL, so we need to check that.
Maybe we should do $keys = $entity_type->getKeys() then do an isset() on the individual keys.

File: src/EntityKeysFieldsTrait.php [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+use Drupal\Core\Entity\ContentEntityTypeInterface;
+use Drupal\Core\Field\BaseFieldDefinition;
+
+/**
+ * Provides base fields for entity keys.
+ */
+trait EntityKeysFieldsTrait {
+
+  /**
+   * Returns some base field definitions.
```
Comment (bojanz) [inline src/EntityKeysFieldsTrait.php]: "Returns the base field definitions for entity keys."

**Comment by dawehner** [comment]
We now ensure that each key exists

File: src/EntityKeysFieldsTrait.php [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        ->setSetting('unsigned', TRUE);
+    }
+
+    if ($entity_type->hasKey('uuid')) {
+      $fields[$entity_type->getKey('uuid')] = BaseFieldDefinition::create('uuid')
+        ->setLabel(t('UUID'))
+        ->setReadOnly(TRUE);
+    }
+
+    if ($entity_type->isRevisionable() && $entity_type->hasKey('revision')) {
```
Comment (bojanz) [inline src/EntityKeysFieldsTrait.php]: isRevisionable() is just a wrapper for hasKey('revision') :)

**Comment by dawehner** [comment]
Ha, good point

**Pull request merged**